### ISSUE TITLE
Issues with whitespace definition

### ISSUE THREAD

**Issue by thobe** [opencypher/openCypher#361]
Neither Java's `Character.isWhitespace(int)`, or `Character.isSpaceChar(int)`, or the unicode [`[:Whitespace:]`](https://unicode.org/cldr/utility/list-unicodeset.jsp?a=%5B%3AWhitespace%3A%5D) specification treats [`\u180E` (MONGOLIAN VOWEL SEPARATOR)](https://unicode.org/cldr/utility/character.jsp?a=180E) as a whitespace.

Yet the openCypher grammar considers this a whitespace character, why?
https://github.com/opencypher/openCypher/blob/346aa0d6ec0d500c84bcc08edb633ba9d8815e73/grammar/basic-grammar.xml#L781

Furthermore the definition of whitespace in the openCypher grammar does not consider [`\u0085` (NEXT LINE)](https://unicode.org/cldr/utility/character.jsp?a=0085) or [`\u00A0` (NBSP)](https://unicode.org/cldr/utility/character.jsp?a=00A0) to be whitespace, while they are part of the unicode [`[:Whitespace:]`](https://unicode.org/cldr/utility/list-unicodeset.jsp?a=%5B%3AWhitespace%3A%5D) specification and considered whitespace by Java's `Character.isSpaceChar(int)`, why?

**Comment by thobe** [comment]
I came across this difference when looking at why the `whitespace` production rules spelled out all whitespace characters individually instead of just referencing the unicode [`[:Whitespace:]`](https://unicode.org/cldr/utility/list-unicodeset.jsp?a=%5B%3AWhitespace%3A%5D) specification. So I investigated the difference.

The conclusion of this exercise is that apart from [`\u0085` (NEXT LINE)](https://unicode.org/cldr/utility/character.jsp?a=0085), the grammar includes all characters of the unicode [`[:Whitespace:]`](https://unicode.org/cldr/utility/list-unicodeset.jsp?a=%5B%3AWhitespace%3A%5D) specification, and additionally includes [`\u001C` (FILE SEPARATOR)](https://unicode.org/cldr/utility/character.jsp?a=001C), [`\u001D` (GROUP SEPARATOR)](https://unicode.org/cldr/utility/character.jsp?a=001D), [`\u001E` (RECORD SEPARATOR)](https://unicode.org/cldr/utility/character.jsp?a=001E), and [`\u001F` (UNIT SEPARATOR)](https://unicode.org/cldr/utility/character.jsp?a=001F).

Tabulating the characters involved:

Code Point | `Character.isWhitespace(...)` | `Character.isSpaceChar(...)` | [`[:Whitespace:]`](https://unicode.org/cldr/utility/list-unicodeset.jsp?a=%5B%3AWhitespace%3A%5D)
-------- | -------- | ------- | -------
[`\u0009`](https://unicode.org/cldr/utility/character.jsp?a=0009) | True | False | True
[`\u000a`](https://unicode.org/cldr/utility/character.jsp?a=000a) | True | False | True
[`\u000b`](https://unicode.org/cldr/utility/character.jsp?a=000b) | True | False | True
[`\u000c`](https://unicode.org/cldr/utility/character.jsp?a=000c) | True | False | True
[`\u000d`](https://unicode.org/cldr/utility/character.jsp?a=000d) | True | False | True
[`\u001c`](https://unicode.org/cldr/utility/character.jsp?a=001c) | True | False | False
[`\u001d`](https://unicode.org/cldr/utility/character.jsp?a=001d) | True | False | False
[`\u001e`](https://unicode.org/cldr/utility/character.jsp?a=001e) | True | False | False
[`\u001f`](https://unicode.org/cldr/utility/character.jsp?a=001f) | True | False | False
[`\u0020`](https://unicode.org/cldr/utility/character.jsp?a=0020) | True | True | True
[`\u0085`](https://unicode.org/cldr/utility/character.jsp?a=0085) | False | False | True
[`\u00a0`](https://unicode.org/cldr/utility/character.jsp?a=00a0) | False | True | True
[`\u1680`](https://unicode.org/cldr/utility/character.jsp?a=1680) | True | True | True
[`\u2000`](https://unicode.org/cldr/utility/character.jsp?a=2000) | True | True | True
[`\u2001`](https://unicode.org/cldr/utility/character.jsp?a=2001) | True | True | True
[`\u2002`](https://unicode.org/cldr/utility/character.jsp?a=2002) | True | True | True
[`\u2003`](https://unicode.org/cldr/utility/character.jsp?a=2003) | True | True | True
[`\u2004`](https://unicode.org/cldr/utility/character.jsp?a=2004) | True | True | True
[`\u2005`](https://unicode.org/cldr/utility/character.jsp?a=2005) | True | True | True
[`\u2006`](https://unicode.org/cldr/utility/character.jsp?a=2006) | True | True | True
[`\u2007`](https://unicode.org/cldr/utility/character.jsp?a=2007) | False | True | True
[`\u2008`](https://unicode.org/cldr/utility/character.jsp?a=2008) | True | True | True
[`\u2009`](https://unicode.org/cldr/utility/character.jsp?a=2009) | True | True | True
[`\u200a`](https://unicode.org/cldr/utility/character.jsp?a=200a) | True | True | True
[`\u2028`](https://unicode.org/cldr/utility/character.jsp?a=2028) | True | True | True
[`\u2029`](https://unicode.org/cldr/utility/character.jsp?a=2029) | True | True | True
[`\u202f`](https://unicode.org/cldr/utility/character.jsp?a=202f) | False | True | True
[`\u205f`](https://unicode.org/cldr/utility/character.jsp?a=205f) | True | True | True
[`\u3000`](https://unicode.org/cldr/utility/character.jsp?a=3000) | True | True | True

**Comment by thobe** [comment]
If we agree to use the unicode [`[:White_Space:]`](https://unicode.org/cldr/utility/list-unicodeset.jsp?a=%5B%3AWhite_Space%3A%5D) specification, we could define whitespace as:

```xml
<production name="whitespace">
  <alt>
    <character set="White_Space"/>
    <character set="FS"/>
    <character set="GS"/>
    <character set="RS"/>
    <character set="US"/>
  </alt>
</production>
```

**Comment by thobe** [comment]
Looking at commit history, it appears as if at some point Java's `Character.isWhitespace(int)` treated [`\u180E` (MONGOLIAN VOWEL SEPARATOR)](https://unicode.org/cldr/utility/character.jsp?a=180E) as a whitespace. At least that is what the code comments say.

**Comment by Mats-SX** [comment]
I think it makes good sense to stick with Unicode here. Do we even need the special additions of FS, GS, RS and US?

**Comment by thobe** [comment]
The `FILE SEPARATOR`, `GROUP SEPARATOR`, `RECORD SEPARATOR`, and `UNIT SEPARATOR` have been explicitly treated as whitespace by Java since forever, and thus by the Neo4j Cypher parser.

They are likely to not occur in Cypher queries. I'd say it's harmless to either include or exclude them.

**Comment by Mats-SX** [comment]
I agree. I would lean towards going with Unicode rather than Java (and abandon Cypher's implementation history), but I don't feel strongly about it. I wonder if any of the two alternatives makes a difference for implementability? I doubt it.

**Comment by hvub** [comment]
See 530
